core/src/main/java/org/infinispan/marshall/core/BytesObjectOutput.java

```diff
@@ -154,68 +154,57 @@ private boolean isAscii(String s, int len) {
 
    @Override
    public void writeUTF(String s) {
-      int startPos = skipIntSize();
+      int strlen = s.length();
+      int startPos = pos;
+      // First optimize for 1 - 127 case
+      ensureCapacity(strlen + 4);
+      // Note this will be overwritten if not all 1 - 127 characters below
+      writeIntDirect(strlen, startPos);
+
       int localPos = pos; /* avoid getfield opcode */
       byte[] localBuf = bytes; /* avoid getfield opcode */
 
-      int strlen = s.length();
-      int c = 0;
+      localPos += 4;
 
-      int i=0;
-      for (i=0; i<strlen; i++) {
+      int c;
+      int i;
+      for (i = 0; i < strlen; i++) {
          c = s.charAt(i);
-         if (!((c >= 0x0001) && (c <= 0x007F))) break;
+         if (c > 127) break;
 
-         if(localPos == bytes.length) {
-            pos = localPos;
-            ensureCapacity(1);
-            localBuf = bytes;
-         }
          localBuf[localPos++] = (byte) c;
       }
 
-      for (;i < strlen; i++){
+      // Means we completed with all latin characters
+      if (i == strlen) {
+         pos = localPos;
```

Comment:
Actually this line needs to be outside of the if, otherwise the ensureCapacity a couple lines below may not resize properly.